File: src/primitives/Hash/Hash.ts
/**
 * Hash Types and Utilities
 *
 * 32-byte hash values with encoding/decoding and comparison utilities.
 * Functions grouped via 'export * as Hash' in index.ts.
 *
 * @example
 * ```typescript
 * import { Hash } from './hash.js';
 *
 * // Create hash from hex
 * const hash = Hash.fromHex('0x1234...');
 *
 * // Convert to hex
 * const hex = Hash.toHex.call(hash);
 *
 * // Compare hashes
 * const same = Hash.equals.call(hash1, hash2);
 *
 * // Hash data with keccak256
 * const digest = Hash.keccak256(data);
 * ```
 */

import { keccak_256 } from "@noble/hashes/sha3.js";

// ============================================================================
// Core Types
// ============================================================================

/**
 * Brand symbol for type safety
 */
export const hashSymbol = Symbol("Hash");

/**
 * Hash size in bytes (32 bytes = 256 bits)
 */
export const SIZE = 32;

/**
 * Zero hash constant (32 zero bytes)
 */
export const ZERO = new Uint8Array(SIZE) as Hash;

// ============================================================================
// Creation Functions
// ============================================================================

/**
 * Create Hash from hex string
 *
 * @param hex - Hex string with optional 0x prefix
 * @returns Hash bytes
 * @throws If hex is invalid or wrong length
 *
 * @example
 * ```typescript
 * const hash = Hash.fromHex('0x1234...');
 * const hash2 = Hash.fromHex('1234...'); // 0x prefix optional
 * ```
 */
export function fromHex(hex: string): Hash {
	const normalized = hex.startsWith("0x") ? hex.slice(2) : hex;
	if (normalized.length !== SIZE * 2) {
		throw new Error(
			`Hash hex must be ${SIZE * 2} characters, got ${normalized.length}`,
		);
	}
	if (!/^[0-9a-fA-F]+$/.test(normalized)) {
		throw new Error("Invalid hex string");
	}
	const bytes = new Uint8Array(SIZE);
	for (let i = 0; i < SIZE; i++) {
		bytes[i] = Number.parseInt(normalized.slice(i * 2, i * 2 + 2), 16);
	}
	return bytes as Hash;
}

/**
 * Create Hash from raw bytes
 *
 * @param bytes - Raw bytes (must be 32 bytes)
 * @returns Hash bytes
 * @throws If bytes is wrong length
 *
 * @example
 * ```typescript
 * const hash = Hash.fromBytes(new Uint8Array(32));
 * ```
 */
export function fromBytes(bytes: Uint8Array): Hash {
	if (bytes.length !== SIZE) {
		throw new Error(`Hash must be ${SIZE} bytes, got ${bytes.length}`);
	}
	return new Uint8Array(bytes) as Hash;
}

/**
 * Create Hash from string or bytes
 *
 * @param value - Hex string with optional 0x prefix or Uint8Array
 * @returns Hash bytes
 *
 * @example
 * ```typescript
 * const hash = Hash.from('0x1234...');
 * const hash2 = Hash.from(new Uint8Array(32));
 * ```
 */
export function from(value: string | Uint8Array): Hash {
	if (typeof value === "string") {
		return fromHex(value);
	}
	return fromBytes(value);
}

// ============================================================================
// Conversion Functions (Internal implementations with this: pattern)
// ============================================================================

/**
 * Convert Hash to hex string (internal)
 *
 * @returns Hex string with 0x prefix
 *
 * @example
 * ```typescript
 * const hex = Hash._toHex.call(hash);
 * // "0x1234..."
 * ```
 */
export function _toHex(this: Hash): string {
	return `0x${Array.from(this, (byte) => byte.toString(16).padStart(2, "0")).join("")}`;
}

/**
 * Convert Hash to raw bytes (internal)
 *
 * @returns Copy of hash bytes
 *
 * @example
 * ```typescript
 * const bytes = Hash._toBytes.call(hash);
 * ```
 */
export function _toBytes(this: Hash): Uint8Array {
	return new Uint8Array(this);
}

/**
 * Convert Hash to string (internal, alias for toHex)
 *
 * @returns Hex string with 0x prefix
 *
 * @example
 * ```typescript
 * const str = Hash._toString.call(hash);
 * ```
 */
export function _toString(this: Hash): string {
	return `0x${Array.from(this, (byte) => byte.toString(16).padStart(2, "0")).join("")}`;
}

// ============================================================================
// Comparison Functions (Internal implementations with this: pattern)
// ============================================================================

/**
 * Compare this hash with another for equality (internal)
 *
 * Uses constant-time comparison to prevent timing attacks.
 *
 * @param other - Hash to compare with
 * @returns True if hashes are equal
 *
 * @example
 * ```typescript
 * const same = Hash._equals.call(hash1, hash2);
 * ```
 */
export function _equals(this: Hash, other: Hash): boolean {
	if (this.length !== other.length) {
		return false;
	}
	let result = 0;
	for (let i = 0; i < this.length; i++) {
		result |= this[i]! ^ other[i]!;
	}
	return result === 0;
}

/**
 * Check if this hash is zero hash (internal)
 *
 * @returns True if hash is all zeros
 *
 * @example
 * ```typescript
 * const isZero = Hash._isZero.call(hash);
 * ```
 */
export function _isZero(this: Hash): boolean {
	if (this.length !== ZERO.length) {
		return false;
	}
	let result = 0;
	for (let i = 0; i < this.length; i++) {
		result |= this[i]! ^ ZERO[i]!;
	}
	return result === 0;
}

// ============================================================================
// Validation Functions
// ============================================================================

/**
 * Check if value is a valid Hash
 *
 * @param value - Value to check
 * @returns True if value is Hash type
 *
 * @example
 * ```typescript
 * if (Hash.isHash(value)) {
 *   // value is Hash
 * }
 * ```
 */
export function isHash(value: unknown): value is Hash {
	return value instanceof Uint8Array && value.length === SIZE;
}

/**
 * Validate hex string is valid hash format
 *
 * @param hex - Hex string to validate
 * @returns True if valid hash hex format
 *
 * @example
 * ```typescript
 * if (Hash.isValidHex('0x1234...')) {
 *   const hash = Hash.fromHex('0x1234...');
 * }
 * ```
 */
export function isValidHex(hex: string): boolean {
	const normalized = hex.startsWith("0x") ? hex.slice(2) : hex;
	return normalized.length === SIZE * 2 && /^[0-9a-fA-F]+$/.test(normalized);
}

/**
 * Assert value is a Hash, throws if not
 *
 * @param value - Value to assert
 * @param message - Optional error message
 * @throws If value is not a Hash
 *
 * @example
 * ```typescript
 * Hash.assert(value); // throws if not Hash
 * ```
 */
export function assert(
	value: unknown,
	message?: string,
): asserts value is Hash {
	if (!isHash(value)) {
		throw new Error(message ?? "Value is not a Hash");
	}
}

// ============================================================================
// Hashing Functions
// ============================================================================

/**
 * Hash data with Keccak-256
 *
 * @param data - Data to hash
 * @returns 32-byte hash
 *
 * @example
 * ```typescript
 * const hash = Hash.keccak256(data);
 * ```
 */
export function keccak256(data: Uint8Array): Hash {
	return keccak_256(data) as Hash;
}

/**
 * Hash string with Keccak-256
 *
 * @param str - String to hash (UTF-8 encoded)
 * @returns 32-byte hash
 *
 * @example
 * ```typescript
 * const hash = Hash.keccak256String('hello');
 * ```
 */
export function keccak256String(str: string): Hash {
	const encoder = new TextEncoder();
	return keccak256(encoder.encode(str));
}

/**
 * Hash hex string with Keccak-256
 *
 * @param hex - Hex string to hash (with or without 0x prefix)
 * @returns 32-byte hash
 *
 * @example
 * ```typescript
 * const hash = Hash.keccak256Hex('0x1234...');
 * ```
 */
export function keccak256Hex(hex: string): Hash {
	const normalized = hex.startsWith("0x") ? hex.slice(2) : hex;
	if (normalized.length % 2 !== 0) {
		throw new Error("Hex string must have even length");
	}
	const bytes = new Uint8Array(normalized.length / 2);
	for (let i = 0; i < bytes.length; i++) {
		bytes[i] = Number.parseInt(normalized.slice(i * 2, i * 2 + 2), 16);
	}
	return keccak256(bytes);
}

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Generate random hash
 *
 * @returns Random 32-byte hash
 *
 * @example
 * ```typescript
 * const hash = Hash.random();
 * ```
 */
export function random(): Hash {
	if (typeof crypto !== "undefined" && crypto.getRandomValues) {
		const bytes = new Uint8Array(SIZE);
		crypto.getRandomValues(bytes);
		return bytes as Hash;
	}
	throw new Error("crypto.getRandomValues not available");
}

/**
 * Clone this hash (internal)
 *
 * @returns New hash with same value
 *
 * @example
 * ```typescript
 * const copy = Hash._clone.call(hash);
 * ```
 */
export function _clone(this: Hash): Hash {
	return new Uint8Array(this) as Hash;
}

/**
 * Get slice of this hash (internal)
 *
 * @param start - Start index (inclusive)
 * @param end - End index (exclusive)
 * @returns Slice of hash bytes
 *
 * @example
 * ```typescript
 * const selector = Hash._slice.call(hash, 0, 4);
 * ```
 */
export function _slice(this: Hash, start?: number, end?: number): Uint8Array {
	return this.slice(start, end);
}

/**
 * Format this hash for display (internal, truncated)
 *
 * @param prefixLength - Number of chars to show at start (default 6)
 * @param suffixLength - Number of chars to show at end (default 4)
 * @returns Formatted string like "0x1234...5678"
 *
 * @example
 * ```typescript
 * const display = Hash._format.call(hash);
 * // "0x1234...5678"
 * ```
 */
export function _format(
	this: Hash,
	prefixLength = 6,
	suffixLength = 4,
): string {
	const hex = `0x${Array.from(this, (byte) => byte.toString(16).padStart(2, "0")).join("")}`;
	if (hex.length <= prefixLength + suffixLength + 2) {
		return hex;
	}
	return `${hex.slice(0, prefixLength + 2)}...${hex.slice(-suffixLength)}`;
}

// ============================================================================
// Type Definition
// ============================================================================

/**
 * Hash type: 32-byte hash value
 *
 * Branded type for type safety.
 */
export type Hash = Uint8Array & { __brand: typeof hashSymbol };

// ============================================================================
// Public Wrapper Functions (namespace+type overloading pattern)
// ============================================================================

/**
 * Convert Hash to hex string
 *
 * @param value - Value to convert to Hash first
 * @returns Hex string with 0x prefix
 *
 * @example
 * ```typescript
 * const hex = Hash.toHex("0x1234...");
 * // "0x1234..."
 * ```
 */
export function toHex(value: string | Uint8Array): string {
	return _toHex.call(from(value));
}

/**
 * Convert Hash to raw bytes
 *
 * @param value - Value to convert to Hash first
 * @returns Copy of hash bytes
 *
 * @example
 * ```typescript
 * const bytes = Hash.toBytes("0x1234...");
 * ```
 */
export function toBytes(value: string | Uint8Array): Uint8Array {
	return _toBytes.call(from(value));
}

/**
 * Convert Hash to string
 *
 * @param value - Value to convert to Hash first
 * @returns Hex string with 0x prefix
 *
 * @example
 * ```typescript
 * const str = Hash.toString("0x1234...");
 * ```
 */
export function toString(value: string | Uint8Array): string {
	return _toString.call(from(value));
}

/**
 * Compare two hashes for equality
 *
 * @param value - First hash value
 * @param other - Hash to compare with
 * @returns True if hashes are equal
 *
 * @example
 * ```typescript
 * const same = Hash.equals("0x1234...", hash2);
 * ```
 */
export function equals(value: string | Uint8Array, other: Hash): boolean {
	return _equals.call(from(value), other);
}

/**
 * Check if hash is zero hash
 *
 * @param value - Value to check
 * @returns True if hash is all zeros
 *
 * @example
 * ```typescript
 * const isZero = Hash.isZero("0x00...");
 * ```
 */
export function isZero(value: string | Uint8Array): boolean {
	return _isZero.call(from(value));
}

/**
 * Clone hash
 *
 * @param value - Value to clone
 * @returns New hash with same value
 *
 * @example
 * ```typescript
 * const copy = Hash.clone("0x1234...");
 * ```
 */
export function clone(value: string | Uint8Array): Hash {
	return _clone.call(from(value));
}

/**
 * Get slice of hash
 *
 * @param value - Hash value to slice
 * @param start - Start index (inclusive)
 * @param end - End index (exclusive)
 * @returns Slice of hash bytes
 *
 * @example
 * ```typescript
 * const selector = Hash.slice("0x1234...", 0, 4);
 * ```
 */
export function slice(
	value: string | Uint8Array,
	start?: number,
	end?: number,
): Uint8Array {
	return _slice.call(from(value), start, end);
}

/**
 * Format hash for display (truncated)
 *
 * @param value - Hash value to format
 * @param prefixLength - Number of chars to show at start (default 6)
 * @param suffixLength - Number of chars to show at end (default 4)
 * @returns Formatted string like "0x1234...5678"
 *
 * @example
 * ```typescript
 * const display = Hash.format("0x1234...");
 * // "0x1234...5678"
 * ```
 */
export function format(
	value: string | Uint8Array,
	prefixLength = 6,
	suffixLength = 4,
): string {
	return _format.call(from(value), prefixLength, suffixLength);
}

File: packages/voltaire-zig/src/primitives/Hash/Hash.zig
//! Hash - 32-byte cryptographic hash type for Ethereum
//!
//! This module provides a strongly-typed 32-byte hash implementation
//! that is used throughout Ethereum for various purposes including
//! transaction hashes, block hashes, state roots, and keccak256 digests.
//!
//! ## Usage
//! ```zig
//! const Hash = @import("primitives").Hash;
//!
//! // From hex string
//! const hash = try Hash.fromHex("0x1234...");
//!
//! // From bytes
//! const hash2 = Hash.fromBytes(&bytes);
//!
//! // Keccak256 hashing
//! const hash3 = Hash.keccak256(&data);
//! ```

const std = @import("std");
const Hex = @import("../Hex/Hex.zig");
const crypto = @import("crypto");

/// Hash size in bytes (32 bytes = 256 bits)
pub const SIZE = 32;

/// Hash type - 32 bytes
pub const Hash = [SIZE]u8;

/// Zero hash constant
pub const ZERO: Hash = [_]u8{0} ** SIZE;

// ============================================================================
// Constructors
// ============================================================================

/// Create Hash from bytes. Input must be exactly 32 bytes.
pub fn fromBytes(bytes: []const u8) Hash {
    std.debug.assert(bytes.len == SIZE);
    var result: Hash = undefined;
    @memcpy(&result, bytes);
    return result;
}

test "fromBytes - valid 32 bytes" {
    const bytes = [_]u8{1} ** SIZE;
    const hash = fromBytes(&bytes);
    try std.testing.expectEqual(SIZE, hash.len);
    try std.testing.expectEqual(@as(u8, 1), hash[0]);
    try std.testing.expectEqual(@as(u8, 1), hash[31]);
}

test "fromBytes - creates independent copy" {
    var bytes = [_]u8{42} ** SIZE;
    const hash = fromBytes(&bytes);
    bytes[0] = 99;
    try std.testing.expectEqual(@as(u8, 42), hash[0]);
}

/// Create Hash from hex string (with or without 0x prefix).
/// Returns error if hex is invalid or not 32 bytes.
pub fn fromHex(hex: []const u8) !Hash {
    const bytes = try Hex.fromHex(std.testing.allocator, hex);
    defer std.testing.allocator.free(bytes);

    if (bytes.len != SIZE) {
        return error.InvalidHashLength;
    }

    return fromBytes(bytes);
}

test "fromHex - with 0x prefix" {
    const hex = "0x" ++ ("ab" ** SIZE);
    const hash = try fromHex(hex);
    try std.testing.expectEqual(@as(u8, 0xab), hash[0]);
    try std.testing.expectEqual(@as(u8, 0xab), hash[31]);
}

test "fromHex - without 0x prefix" {
    const hex = "cd" ** SIZE;
    const hash = try fromHex(hex);
    try std.testing.expectEqual(@as(u8, 0xcd), hash[0]);
}

test "fromHex - invalid length" {
    const hex = "0xaabb";
    try std.testing.expectError(error.InvalidHashLength, fromHex(hex));
}

test "fromHex - invalid hex chars" {
    const hex = "0x" ++ ("zz" ** SIZE);
    try std.testing.expectError(error.InvalidHexCharacter, fromHex(hex));
}

/// Generic constructor - accepts bytes or hex string
pub fn from(value: anytype) !Hash {
    const T = @TypeOf(value);
    if (T == Hash) return value;

    if (T == []const u8 or T == []u8) {
        // Check if it's hex (starts with 0x or all hex chars)
        if (value.len > 2 and value[0] == '0' and value[1] == 'x') {
            return fromHex(value);
        }
        // Check if it's exactly 32 bytes (raw)
        if (value.len == SIZE) {
            return fromBytes(value);
        }
        // Try as hex without 0x
        if (value.len == SIZE * 2) {
            return fromHex(value);
        }
        return error.InvalidHashInput;
    }

    @compileError("Unsupported type for Hash.from: " ++ @typeName(T));
}

test "from - Hash passthrough" {
    const hash1: Hash = ZERO;
    const hash2 = try from(hash1);
    try std.testing.expectEqualSlices(u8, &hash1, &hash2);
}

test "from - raw bytes" {
    const bytes = [_]u8{42} ** SIZE;
    const hash = try from(bytes[0..]);
    try std.testing.expectEqual(@as(u8, 42), hash[0]);
}

test "from - hex with 0x" {
    const hex = "0x" ++ ("ef" ** SIZE);
    const hash = try from(hex);
    try std.testing.expectEqual(@as(u8, 0xef), hash[0]);
}

test "from - hex without 0x" {
    const hex = "12" ** SIZE;
    const hash = try from(hex);
    try std.testing.expectEqual(@as(u8, 0x12), hash[0]);
}

// ============================================================================
// Converters
// ============================================================================

/// Convert Hash to bytes slice
pub fn toBytes(hash: *const Hash) []const u8 {
    return hash[0..];
}

test "toBytes - returns correct slice" {
    var hash: Hash = undefined;
    @memset(&hash, 0xaa);
    const bytes = toBytes(&hash);
    try std.testing.expectEqual(SIZE, bytes.len);
    try std.testing.expectEqual(@as(u8, 0xaa), bytes[0]);
    try std.testing.expectEqual(@as(u8, 0xaa), bytes[31]);
}

/// Convert Hash to hex string with 0x prefix
pub fn toHex(hash: *const Hash, allocator: std.mem.Allocator) ![]const u8 {
    return try Hex.toHex(allocator, hash[0..]);
}

test "toHex - with 0x prefix" {
    var hash: Hash = undefined;
    @memset(&hash, 0xff);
    const hex = try toHex(&hash, std.testing.allocator);
    defer std.testing.allocator.free(hex);

    try std.testing.expectEqual(2 + SIZE * 2, hex.len);
    try std.testing.expect(std.mem.startsWith(u8, hex, "0x"));
    try std.testing.expect(std.mem.eql(u8, hex, "0x" ++ ("ff" ** SIZE)));
}

/// Alias for toHex for string representation
pub fn toString(hash: *const Hash, allocator: std.mem.Allocator) ![]const u8 {
    return toHex(hash, allocator);
}

test "toString - same as toHex" {
    const hash: Hash = ZERO;
    const str = try toString(&hash, std.testing.allocator);
    defer std.testing.allocator.free(str);
    const hex_str = try toHex(&hash, std.testing.allocator);
    defer std.testing.allocator.free(hex_str);
    try std.testing.expect(std.mem.eql(u8, str, hex_str));
}

// ============================================================================
// Validation
// ============================================================================

/// Compare two hashes for equality
pub fn equals(a: *const Hash, b: *const Hash) bool {
    return std.mem.eql(u8, a[0..], b[0..]);
}

test "equals - same hash" {
    const hash: Hash = [_]u8{42} ** SIZE;
    try std.testing.expect(equals(&hash, &hash));
}

test "equals - identical hashes" {
    const hash1: Hash = [_]u8{99} ** SIZE;
    const hash2: Hash = [_]u8{99} ** SIZE;
    try std.testing.expect(equals(&hash1, &hash2));
}

test "equals - different hashes" {
    const hash1: Hash = [_]u8{1} ** SIZE;
    const hash2: Hash = [_]u8{2} ** SIZE;
    try std.testing.expect(!equals(&hash1, &hash2));
}

/// Check if hash is all zeros
pub fn isZero(hash: *const Hash) bool {
    return equals(hash, &ZERO);
}

test "isZero - zero hash" {
    const hash: Hash = ZERO;
    try std.testing.expect(isZero(&hash));
}

test "isZero - non-zero hash" {
    var hash: Hash = ZERO;
    hash[0] = 1;
    try std.testing.expect(!isZero(&hash));
}

/// Assert that value is a valid hash (for runtime checks)
pub fn assert(hash: *const Hash) void {
    // Hash is always valid if it's the right type
    // This function exists for API parity
    _ = hash;
}

test "assert - always succeeds for Hash type" {
    const hash: Hash = ZERO;
    assert(&hash);
}

/// Check if a hex string is valid hash format
pub fn isValidHex(hex: []const u8) bool {
    // Must be 64 or 66 chars (with/without 0x)
    if (hex.len != SIZE * 2 and hex.len != SIZE * 2 + 2) {
        return false;
    }

    var start: usize = 0;
    if (hex.len == SIZE * 2 + 2) {
        if (hex[0] != '0' or hex[1] != 'x') return false;
        start = 2;
    }

    // Check all chars are hex
    for (hex[start..]) |c| {
        if (!std.ascii.isHex(c)) return false;
    }

    return true;
}

test "isValidHex - valid with 0x" {
    const hex = "0x" ++ ("ab" ** SIZE);
    try std.testing.expect(isValidHex(hex));
}

test "isValidHex - valid without 0x" {
    const hex = "cd" ** SIZE;
    try std.testing.expect(isValidHex(hex));
}

test "isValidHex - invalid length" {
    try std.testing.expect(!isValidHex("0xaabb"));
}

test "isValidHex - invalid chars" {
    const hex = "0x" ++ ("zz" ** SIZE);
    try std.testing.expect(!isValidHex(hex));
}

test "isValidHex - no 0x prefix but wrong length" {
    try std.testing.expect(!isValidHex("abcd"));
}

/// Type guard - check if value is a Hash
pub fn isHash(value: anytype) bool {
    const T = @TypeOf(value);
    return T == Hash or T == *const Hash or T == *Hash;
}

test "isHash - Hash type" {
    const hash: Hash = ZERO;
    try std.testing.expect(isHash(hash));
}

test "isHash - pointer types" {
    const hash: Hash = ZERO;
    try std.testing.expect(isHash(&hash));
}

test "isHash - non-Hash type" {
    const bytes = [_]u8{0} ** 16;
    try std.testing.expect(!isHash(bytes));
}

// ============================================================================
// Manipulation
// ============================================================================

/// Create a copy of the hash
pub fn clone(hash: *const Hash) Hash {
    var result: Hash = undefined;
    @memcpy(&result, hash);
    return result;
}

test "clone - creates independent copy" {
    var original: Hash = [_]u8{42} ** SIZE;
    var copy = clone(&original);
    try std.testing.expect(equals(&original, &copy));

    // Modify original
    original[0] = 99;
    try std.testing.expect(!equals(&original, &copy));
    try std.testing.expectEqual(@as(u8, 42), copy[0]);
}

/// Slice hash bytes (returns a new Hash only if slice is 32 bytes)
pub fn slice(hash: *const Hash, start: usize, end: usize) []const u8 {
    return hash[start..end];
}

test "slice - partial slice" {
    var hash: Hash = undefined;
    for (0..SIZE) |i| {
        hash[i] = @intCast(i);
    }
    const s = slice(&hash, 0, 10);
    try std.testing.expectEqual(@as(usize, 10), s.len);
    try std.testing.expectEqual(@as(u8, 0), s[0]);
    try std.testing.expectEqual(@as(u8, 9), s[9]);
}

test "slice - full slice" {
    const hash: Hash = [_]u8{77} ** SIZE;
    const s = slice(&hash, 0, SIZE);
    try std.testing.expectEqual(SIZE, s.len);
    try std.testing.expectEqualSlices(u8, &hash, s);
}

/// Format hash for display (returns hex string)
pub fn format(hash: *const Hash, allocator: std.mem.Allocator) ![]const u8 {
    return toHex(hash, allocator);
}

test "format - returns hex string" {
    const hash: Hash = ZERO;
    const formatted = try format(&hash, std.testing.allocator);
    defer std.testing.allocator.free(formatted);
    try std.testing.expect(std.mem.startsWith(u8, formatted, "0x"));
    try std.testing.expectEqual(2 + SIZE * 2, formatted.len);
}

/// Generate random hash (for testing)
pub fn random() Hash {
    var hash: Hash = undefined;
    std.crypto.random.bytes(&hash);
    return hash;
}

test "random - generates different hashes" {
    const hash1 = random();
    const hash2 = random();
    // Astronomically unlikely to be equal
    try std.testing.expect(!equals(&hash1, &hash2));
}

test "random - correct size" {
    const hash = random();
    try std.testing.expectEqual(SIZE, hash.len);
}

// ============================================================================
// Cryptographic Operations
// ============================================================================

/// Compute Keccak256 hash of data
pub fn keccak256(data: []const u8) Hash {
    var result: Hash = undefined;
    crypto.Keccak256.hash(data, &result);
    return result;
}

test "keccak256 - empty data" {
    const hash = keccak256("");
    // Known Keccak256 of empty string
    const expected_hex = "0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470";
    const expected = try fromHex(expected_hex);
    try std.testing.expect(equals(&hash, &expected));
}

test "keccak256 - known vector" {
    const hash = keccak256("abc");
    // Known Keccak256("abc")
    const expected_hex = "0x4e03657aea45a94fc7d47ba826c8d667c0d1e6e33a64a036ec44f58fa12d6c45";
    const expected = try fromHex(expected_hex);
    try std.testing.expect(equals(&hash, &expected));
}

/// Compute Keccak256 hash of string
pub fn keccak256String(str: []const u8) Hash {
    return keccak256(str);
}

test "keccak256String - same as keccak256" {
    const str = "test string";
    const hash1 = keccak256String(str);
    const hash2 = keccak256(str);
    try std.testing.expect(equals(&hash1, &hash2));
}

/// Compute Keccak256 hash of hex string
pub fn keccak256Hex(hex: []const u8) !Hash {
    const bytes = try Hex.toBytes(hex, std.testing.allocator);
    defer std.testing.allocator.free(bytes);
    return keccak256(bytes);
}

test "keccak256Hex - valid hex" {
    const hex = "0xaabbccdd";
    const hash = try keccak256Hex(hex);

    // Verify by converting hex to bytes and hashing
    const bytes = try Hex.toBytes(hex, std.testing.allocator);
    defer std.testing.allocator.free(bytes);
    const expected = keccak256(bytes);

    try std.testing.expect(equals(&hash, &expected));
}

// ============================================================================
// Advanced Operations
// ============================================================================

/// Concatenate multiple hashes and hash the result
pub fn concat(allocator: std.mem.Allocator, hashes: []const Hash) !Hash {
    // Calculate total size
    const total_size = hashes.len * SIZE;
    const buffer = try allocator.alloc(u8, total_size);
    defer allocator.free(buffer);

    // Copy all hashes into buffer
    for (hashes, 0..) |hash, i| {
        @memcpy(buffer[i * SIZE .. (i + 1) * SIZE], &hash);
    }

    // Hash the concatenation
    return keccak256(buffer);
}

test "concat - two hashes" {
    const hash1: Hash = [_]u8{0xaa} ** SIZE;
    const hash2: Hash = [_]u8{0xbb} ** SIZE;
    const hashes = [_]Hash{ hash1, hash2 };

    const result = try concat(std.testing.allocator, &hashes);

    // Verify by manually concatenating
    var buffer: [SIZE * 2]u8 = undefined;
    @memcpy(buffer[0..SIZE], &hash1);
    @memcpy(buffer[SIZE .. SIZE * 2], &hash2);
    const expected = keccak256(&buffer);

    try std.testing.expect(equals(&result, &expected));
}

test "concat - single hash" {
    const hash1: Hash = [_]u8{0xcc} ** SIZE;
    const hashes = [_]Hash{hash1};
    const result = try concat(std.testing.allocator, &hashes);
    const expected = keccak256(&hash1);
    try std.testing.expect(equals(&result, &expected));
}

test "concat - empty array returns hash of empty data" {
    const hashes: []const Hash = &[_]Hash{};
    const result = try concat(std.testing.allocator, hashes);
    const expected = keccak256("");
    try std.testing.expect(equals(&result, &expected));
}

/// Compute Merkle root of hash array
pub fn merkleRoot(allocator: std.mem.Allocator, hashes: []const Hash) !Hash {
    if (hashes.len == 0) {
        return ZERO;
    }
    if (hashes.len == 1) {
        return hashes[0];
    }

    // Create working buffer for current level
    var current = try allocator.alloc(Hash, hashes.len);
    defer allocator.free(current);
    @memcpy(current, hashes);

    var current_len = hashes.len;

    // Build tree bottom-up
    while (current_len > 1) {
        const next_len = (current_len + 1) / 2;
        var next = try allocator.alloc(Hash, next_len);
        defer allocator.free(next);

        var i: usize = 0;
        while (i < current_len) : (i += 2) {
            if (i + 1 < current_len) {
                // Pair exists
                const pair = [_]Hash{ current[i], current[i + 1] };
                next[i / 2] = try concat(allocator, &pair);
            } else {
                // Odd one out - hash with itself
                const pair = [_]Hash{ current[i], current[i] };
                next[i / 2] = try concat(allocator, &pair);
            }
        }

        // Swap buffers
        const temp = current;
        current = next;
        current_len = next_len;
        allocator.free(temp);

        // Re-allocate for next iteration if needed
        if (current_len > 1) {
            const new_current = try allocator.alloc(Hash, current_len);
            @memcpy(new_current, current[0..current_len]);
            allocator.free(current);
            current = new_current;
        }
    }

    return current[0];
}

test "merkleRoot - empty array" {
    const hashes: []const Hash = &[_]Hash{};
    const root = try merkleRoot(std.testing.allocator, hashes);
    try std.testing.expect(equals(&root, &ZERO));
}

test "merkleRoot - single hash" {
    const hash: Hash = [_]u8{0xaa} ** SIZE;
    const hashes = [_]Hash{hash};
    const root = try merkleRoot(std.testing.allocator, &hashes);
    try std.testing.expect(equals(&root, &hash));
}

test "merkleRoot - two hashes" {
    const hash1: Hash = [_]u8{0xaa} ** SIZE;
    const hash2: Hash = [_]u8{0xbb} ** SIZE;
    const hashes = [_]Hash{ hash1, hash2 };

    const root = try merkleRoot(std.testing.allocator, &hashes);

    // Root should be concat(hash1, hash2)
    const expected = try concat(std.testing.allocator, &hashes);
    try std.testing.expect(equals(&root, &expected));
}

test "merkleRoot - three hashes (odd)" {
    const hash1: Hash = [_]u8{0x11} ** SIZE;
    const hash2: Hash = [_]u8{0x22} ** SIZE;
    const hash3: Hash = [_]u8{0x33} ** SIZE;
    const hashes = [_]Hash{ hash1, hash2, hash3 };

    const root = try merkleRoot(std.testing.allocator, &hashes);

    // Manually compute expected
    // Level 1: h12 = concat(h1, h2), h33 = concat(h3, h3)
    const pair12 = [_]Hash{ hash1, hash2 };
    const h12 = try concat(std.testing.allocator, &pair12);
    const pair33 = [_]Hash{ hash3, hash3 };
    const h33 = try concat(std.testing.allocator, &pair33);

    // Level 2: root = concat(h12, h33)
    const final_pair = [_]Hash{ h12, h33 };
    const expected = try concat(std.testing.allocator, &final_pair);

    try std.testing.expect(equals(&root, &expected));
}

test "merkleRoot - four hashes (even)" {
    const hash1: Hash = [_]u8{0x11} ** SIZE;
    const hash2: Hash = [_]u8{0x22} ** SIZE;
    const hash3: Hash = [_]u8{0x33} ** SIZE;
    const hash4: Hash = [_]u8{0x44} ** SIZE;
    const hashes = [_]Hash{ hash1, hash2, hash3, hash4 };

    const root = try merkleRoot(std.testing.allocator, &hashes);

    // Manually compute
    const pair12 = [_]Hash{ hash1, hash2 };
    const h12 = try concat(std.testing.allocator, &pair12);
    const pair34 = [_]Hash{ hash3, hash4 };
    const h34 = try concat(std.testing.allocator, &pair34);
    const final_pair = [_]Hash{ h12, h34 };
    const expected = try concat(std.testing.allocator, &final_pair);

    try std.testing.expect(equals(&root, &expected));
}
